### PR TITLE
FIXED: Better species environment handlers, and custom race body temperature

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request

This now fixes hard suit fire suits and overheating issues.

This refactors the handle_environment for most mobs and species. 
Better documentation in the code
Better consistency across mobs.
Basic mobs can now have cold and heat protection form the environmental. (not simple mobs)
Slime's temperature change code has been cleaned up.
Base human default temperature range is a bit more reasonable, (+30, -40 vs +/- 50) you would cook or freeze without taking damage before. (Thoughts)
Split up species handle_environment, it was a single massive proc, now each species has handle_environment, handle_body_temperature, handle_environment_pressure.
Each species can now have their own body temp, and hot/cold damage ranges.
Removed checks that we have an environment in many of the classes, this is handled at the base mob class.

**SPECIES:**
All the species modified will survive the normal range of temperature range on station.
- Ethereals core body temp is hotter and can stand higher body heat, but they take damage in cold areas faster.
- Lizard people are cold blooded and do not stabilize body temperature normally, because of they do not take damage as fast as a human. They still stabilize to the room temp.
- Plasmamen like it cold and can survive colder temps better, but take damage when it is hot faster and they can stabilize body temp slower at times.
- Zombies do not stabilize body heat naturally, they are dead. Heat damage is equal to fire spread temp, and cold temp is equal to min temp to move. They still stabilize to the room temp, like lizards

## Why It's Good For The Game

Code cleanup
Each race now can be special and handle the environment differently.
Human body temp is not normal for everyone now.
You could find a zombie by scanning its body temp.

## Changelog
:cl:
add: Species body temperature handling 
balance: Plasmamen, Lizardpeople, Ethereal have custom body temperatures to deal with
fix: natural body temp stabilization now returns proper values
refactor: clean up of the environment handles for mobs and species
/:cl:

### CODE REVIEW

**Comment by nightred** [comment]
The natural stabilization was off and causing overheating.
You can now survive hard suits, fire suits, bio suits.
Lizards recover from being on fire.

**Comment by itseasytosee** [comment]
So just checking being in a fire suit now no longer makes it worse when you are on fire

File: code/modules/mob/living/carbon/human/species_types/plasmamen.dm [hunk 1]
```diff
@@ -22,6 +22,17 @@
...
 	outfit_important_for_life = /datum/outfit/plasmaman
 
+	// Body temperature for Plasmen is much lower
+	// they can handle colder environments
+	// but they are hurt at hot temps faster as it is harder to hold their form
+	// and they will regain temperature more slowly in some situations
+	bodytemp_normal = (BODYTEMP_NORMAL - 40)
+	bodytemp_autorecovery_min = 2
+	bodytemp_heat_damage_limit = (BODYTEMP_HEAT_DAMAGE_LIMIT - 20) // about 40C
+	// In the event they loss the cold resit trait
```
Comment (ATH1909) [inline code/modules/mob/living/carbon/human/species_types/plasmamen.dm]: *resist

**Comment by ATH1909** [comment]
So, are lizards just gonna die if they don't eat iced tea or something after getting set on fire (and later extinguishing themselves) because their body temperature can't correct itself?

Are cryo stings just gonna become a one click win button against lizardpeople, even if they have coffee or something on their person?

Do zombies reset their internal temperature when they rez, or can you cause a zombie to repeatedly die over and over again if you set it on fire once?

**Comment by nightred** [comment]
Body temp stabilization is only 50% natural, what that means is the turf they are on is what they will stabilize body temp to.

Lizards that are on fire and over 1500 k will drop down to 290 k in about 20 seconds where a human will do the same in half the time. Where this is an issue for the lizards is that if the room is 500 k because of a fire they have no wait to fight that. Now in a human they will still over heat in that room but it will take them twice as long to reach overheat temp, this is because natural stabilization will be trying to bring that down. 

Lizards do not heat up or cool down in hard suits but they maintain where they currently are based on the same magic that makes humans stable in a hard suit.

For a zombie this means that the body will be the same temp as the room and will be fine on revival.
I have no idea how cryo sting will hit a zombie.

**Comment by nightred** [comment]
This has been test merged for several days on Sybil without complain.

**Comment by nightred** [comment]
Ash Lizards were only 0.15 kelvin above lavaland atmos max temp, this gives lizards better lavaland survival.

File: code/modules/mob/living/carbon/carbon_defines.dm [hunk 2]
```diff
@@ -64,3 +64,9 @@
 	var/drunkenness = 0 //Overall drunkenness - check handle_alcohol() in life.dm for effects
 	var/stam_regen_start_time = 0 //used to halt stamina regen temporarily
 	var/stam_paralyzed = FALSE //knocks you down
+
+	// Protection from the elements
+	// This is insulation against increasing or decreasing body temperature
+	// 0 - 1 value which corresponds to the percentage of protection
+	var/heat_protection = 0 // No heat protection
+	var/cold_protection = 0 // No cold protection
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/carbon_defines.dm]: Use doc comments
```
/// This is a doc comment
var/something
```

File: code/modules/mob/living/carbon/human/life.dm [hunk 3]
```diff
@@ -231,11 +242,19 @@
 
 	return thermal_protection_flags
 
-/mob/living/carbon/human/proc/get_cold_protection(temperature)
-	temperature = max(temperature, 2.7) //There is an occasional bug where the temperature is miscalculated in ares with a small amount of gas on them, so this is necessary to ensure that that bug does not affect this calculation. Space's temperature is 2.7K and most suits that are intended to protect against any cold, protect down to 2.0K.
+// Temperature is the temperature you're being exposed to.
+// This returns a 0 - 1 value which corresponds to the percentage of protection
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/life.dm]: Put this as a doc comment on the base proc

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -54,6 +54,13 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	var/datum/action/innate/flight/fly //the actual flying ability given to flying species
 	var/wings_icon = "Angel" //the icon used for the wings
 
+	// Body temperature defaults
+	// this enables different species to have different body temperature settings
+	var/bodytemp_normal = BODYTEMP_NORMAL
+	var/bodytemp_autorecovery_min = BODYTEMP_AUTORECOVERY_MINIMUM
+	var/bodytemp_heat_damage_limit = BODYTEMP_HEAT_DAMAGE_LIMIT
+	var/bodytemp_cold_damage_limit = BODYTEMP_COLD_DAMAGE_LIMIT
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm]: More things needing doc comments

File: code/modules/mob/living/carbon/life.dm [hunk 5]
```diff
@@ -569,18 +569,88 @@ GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put
...
+				natural_change = (thermal_protection + 1) * natural
+			else
+				// Our bodytemp is above normal and sweating, insulation hinders out ability to reduce heat
+				// but will reduce the amount of heat we get from the environment
+				natural_change = (1 / (thermal_protection + 1)) * natural
+
+	// Apply the temperature changes
+	adjust_bodytemperature(natural_change + environment_change)
+
+// Used to stabilize the normal body temperature on living mobs
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/life.dm]:
```suggestion
/// Used to stabilize the normal body temperature on living mobs
```

File: code/modules/mob/living/simple_animal/slime/life.dm [hunk 6]
```diff
@@ -108,14 +108,22 @@
 	AIproc = 0
 
 /mob/living/simple_animal/slime/handle_environment(datum/gas_mixture/environment)
-	if(!environment)
-		return
-
+	// Get the current air temperature
```
Comment (ninjanomnom) [inline code/modules/mob/living/simple_animal/slime/life.dm]:
```suggestion
	/// Get the current air temperature
```

File: code/modules/mob/living/simple_animal/slime/life.dm [hunk 7]
```diff
@@ -108,14 +108,22 @@
...
 	var/loc_temp = get_temperature(environment)
+	// This divisor controls how fast body temperature changes
```
Comment (ninjanomnom) [inline code/modules/mob/living/simple_animal/slime/life.dm]:
```suggestion
	/// This divisor controls how fast body temperature changes
```

File: code/modules/mob/living/simple_animal/slime/life.dm [hunk 8]
```diff
@@ -108,14 +108,22 @@
...
+	// lowering this number changes body temp faster
```
Comment (ninjanomnom) [inline code/modules/mob/living/simple_animal/slime/life.dm]:
```suggestion
	/// lowering this number changes body temp faster
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by nightred** [comment]
Updated the comments and used mode doc comments. does this clean up work for you ninja?

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**